crates/legacy/src/testing/finalization_test.rs

```diff
@@ -0,0 +1,550 @@
+
+        BuilderStateId {
+            parent_commitment: payload_vid_commitment,
+            view: next_view,
+        }
+    }
+}
+
+/// [test_empty_block_rate] is a test to ensure that if we don't have any
+/// transactions being submitted, that the builder will continue it's current
+/// behavior of not proposing empty blocks.
+///
+/// |> Note: this test simulates how consensus interacts with the Builder in a
+/// |> very basic way.  When consensus asks for available blocks, and the
+/// |> Builder returns an error that indicates that it does not have any blocks
+/// |> to propose, consensus will retry a few times before giving up. As a
+/// |> result the number of times that consensus has to ask the Builder for
+/// |> block is an integral part of this test.
+#[async_std::test]
+async fn test_empty_block_rate() {
+    let (proxy_global_state, _, da_proposal_sender, quorum_proposal_sender, _) =
+        setup_builder_for_test(None);
+
+    let mut current_builder_state_id = BuilderStateId::<TestTypes> {
+        parent_commitment: vid_commitment(&[], TEST_NUM_NODES_IN_VID_COMPUTATION),
+        view: ViewNumber::genesis(),
+    };
+
+    for round in 0..10 {
+        let (attempts, available_available_blocks_result) = process_available_blocks_round(
+            &proxy_global_state,
+            current_builder_state_id.clone(),
+            round,
+        )
+        .await;
+
+        assert_eq!(
+            attempts, TEST_NUM_CONSENSUS_RETRIES,
+            "Consensus should retry {TEST_NUM_CONSENSUS_RETRIES} times to get available blocks"
+        );
+        assert!(available_available_blocks_result.is_err());
+
+        current_builder_state_id = progress_round_without_available_block_info(
+            current_builder_state_id,
+            round,
+            &da_proposal_sender,
+            &quorum_proposal_sender,
+        )
+        .await;
+    }
+}
+
+/// [test_eager_block_rate] is a test that ensures that the builder will propose
+/// empty blocks, if consensus indicates a proposal included transactions.
+///
+/// It checks initially that it does not propose any empty blocks in round 0.
+/// It checks that it proposes a block with transactions in round 1, which
+/// gets included by consensus.
+/// It then checks that the next `allow_empty_block_period` rounds return empty
+/// blocks without the need to retry.
+/// It then checks that the remaining round up to 9 will not propose any empty
+/// blocks.
+///
+/// |> Note: this test simulates how consensus interacts with the Builder in a
+/// |> very basic way.  When consensus asks for available blocks, and the
+/// |> Builder returns an error that indicates that it does not have any blocks
+/// |> to propose, consensus will retry a few times before giving up. As a
+/// |> result the number of times that consensus has to ask the Builder for
+/// |> block is an integral part of this test.
+#[async_std::test]
+async fn test_eager_block_rate() {
```

Comment:
Love the structure of this test, it's super clear which stage the current status is in.